File: src/obsidian_mcp/canvas.py
"""Canvas file CRUD for Obsidian .canvas files."""

from __future__ import annotations

import json
import uuid
from pathlib import Path

from mcp.server.fastmcp import FastMCP, Context

from obsidian_mcp.errors import CanvasParseError
from obsidian_mcp.vault import Vault


def _resolve_canvas(vault: Vault, path: str) -> Path:
    if not path.endswith(".canvas"):
        path = path + ".canvas"
    return vault.root / path


def _new_id() -> str:
    return uuid.uuid4().hex[:8]


def read_canvas(vault: Vault, path: str) -> dict:
    canvas_path = _resolve_canvas(vault, path)
    if not canvas_path.exists():
        raise FileNotFoundError(f"Canvas not found: {path}")
    try:
        return json.loads(canvas_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise CanvasParseError(f"Malformed canvas JSON in {path}: {exc}") from exc


def create_canvas(
    vault: Vault,
    path: str,
    nodes: list[dict] | None = None,
    edges: list[dict] | None = None,
) -> str:
    canvas_path = _resolve_canvas(vault, path)
    if canvas_path.exists():
        raise FileExistsError(f"Canvas already exists: {path}")

    canvas_path.parent.mkdir(parents=True, exist_ok=True)
    data = {
        "nodes": nodes or [],
        "edges": edges or [],
    }
    canvas_path.write_text(json.dumps(data, indent=2), encoding="utf-8")
    return str(canvas_path.relative_to(vault.root))


def update_canvas(vault: Vault, path: str, data: dict) -> str:
    canvas_path = _resolve_canvas(vault, path)
    if not canvas_path.exists():
        raise FileNotFoundError(f"Canvas not found: {path}")
    canvas_path.write_text(json.dumps(data, indent=2), encoding="utf-8")
    return str(canvas_path.relative_to(vault.root))


def delete_canvas(vault: Vault, path: str) -> str:
    canvas_path = _resolve_canvas(vault, path)
    if not canvas_path.exists():
        raise FileNotFoundError(f"Canvas not found: {path}")
    rel = str(canvas_path.relative_to(vault.root))
    canvas_path.unlink()
    return rel


def register_tools(mcp: FastMCP) -> None:
    """Register canvas tools with the MCP server."""
    from mcp.server.fastmcp.exceptions import ToolError

    @mcp.tool()
    async def canvas_read(path: str, ctx: Context = None) -> dict:
        """Read and return a .canvas file as JSON."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return read_canvas(vault, path)
        except FileNotFoundError as e:
            raise ToolError(str(e))
        except CanvasParseError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

    @mcp.tool()
    async def canvas_create(
        path: str,
        nodes: list[dict] | None = None,
        edges: list[dict] | None = None,
        ctx: Context = None,
    ) -> str:
        """Create a new .canvas file."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return create_canvas(vault, path, nodes, edges)
        except FileExistsError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

    @mcp.tool()
    async def canvas_update(path: str, data: dict, ctx: Context = None) -> str:
        """Overwrite a .canvas file with new data."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return update_canvas(vault, path, data)
        except FileNotFoundError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

    @mcp.tool()
    async def canvas_delete(path: str, ctx: Context = None) -> str:
        """Delete a .canvas file."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return delete_canvas(vault, path)
        except FileNotFoundError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

File: rust/src/canvas.rs
use std::fs;
use std::path::PathBuf;
use serde_json::{json, Value};

use crate::errors::{ObsidianError, Result};
use crate::vault::Vault;

fn resolve_canvas(vault: &Vault, path: &str) -> PathBuf {
    let path_with_ext = if path.ends_with(".canvas") {
        path.to_string()
    } else {
        format!("{}.canvas", path)
    };
    vault.root.join(path_with_ext)
}

/// Read and return a .canvas file as a JSON Value.
pub fn read_canvas(vault: &Vault, path: &str) -> Result<Value> {
    let canvas_path = resolve_canvas(vault, path);
    if !canvas_path.exists() {
        return Err(ObsidianError::NoteNotFound(format!(
            "Canvas not found: {}",
            path
        )));
    }
    let text = fs::read_to_string(&canvas_path)?;
    let val: Value = serde_json::from_str(&text).map_err(|e| {
        ObsidianError::CanvasParseError(format!("Malformed canvas JSON in {}: {}", path, e))
    })?;
    Ok(val)
}

/// Create a new .canvas file. Returns its relative path.
pub fn create_canvas(
    vault: &Vault,
    path: &str,
    nodes: Option<Value>,
    edges: Option<Value>,
) -> Result<String> {
    let canvas_path = resolve_canvas(vault, path);
    if canvas_path.exists() {
        return Err(ObsidianError::NoteAlreadyExists(format!(
            "Canvas already exists: {}",
            path
        )));
    }
    if let Some(parent) = canvas_path.parent() {
        fs::create_dir_all(parent)?;
    }
    let data = json!({
        "nodes": nodes.unwrap_or_else(|| json!([])),
        "edges": edges.unwrap_or_else(|| json!([])),
    });
    fs::write(&canvas_path, serde_json::to_string_pretty(&data)?.as_bytes())?;
    Ok(vault.relative(&canvas_path))
}

/// Overwrite a .canvas file with new JSON data. Returns its relative path.
pub fn update_canvas(vault: &Vault, path: &str, data: &Value) -> Result<String> {
    let canvas_path = resolve_canvas(vault, path);
    if !canvas_path.exists() {
        return Err(ObsidianError::NoteNotFound(format!(
            "Canvas not found: {}",
            path
        )));
    }
    fs::write(&canvas_path, serde_json::to_string_pretty(data)?.as_bytes())?;
    Ok(vault.relative(&canvas_path))
}

/// Delete a .canvas file. Returns its former relative path.
pub fn delete_canvas(vault: &Vault, path: &str) -> Result<String> {
    let canvas_path = resolve_canvas(vault, path);
    if !canvas_path.exists() {
        return Err(ObsidianError::NoteNotFound(format!(
            "Canvas not found: {}",
            path
        )));
    }
    let rel = vault.relative(&canvas_path);
    fs::remove_file(&canvas_path)?;
    Ok(rel)
}
